tests/integration/general/upgrade.bats

```diff
@@ -0,0 +1,186 @@
+  load_common "git-version.bash"
+  load_common "yq.bash"
+  load_assert
+  load_file
+  load_mock
+
+  env.private
+
+  gitversion.setup_mocks
+
+  # consistent state at start
+  run yq -i '.global.ck8sVersion="v0.41.0"' "${CK8S_CONFIG_PATH}/defaults/common-config.yaml"
+  ck8s ops kubectl sc delete -n kube-system configmap apps-upgrade --ignore-not-found --wait
+  ck8s ops kubectl sc apply -f - <<EOF
+apiVersion: v1
+data:
+  version: v0.41
+kind: ConfigMap
+metadata:
+  name: apps-meta
+  namespace: kube-system
+EOF
+}
+
+teardown() {
+  # reset
+
+  env.teardown
+}
+
+@test "upgrading works" {
+  # test 1, the happy path where everything goes well
+
+  run ck8s version config
+  assert_success
+  assert_output --partial "v0.41"
+
+  gitversion.mock_static "v0.42.0"
+  run ck8s upgrade sc "v0.42" prepare
+  assert_success
+  assert_output --partial "locked for upgrade"
+
+  gitversion.mock_static "v0.42.0"
+  run ck8s upgrade sc "v0.42" apply
+  assert_success
+
+  run ck8s apply sc
+  assert_success
+
+  run ck8s version sc
+  assert_success
+  assert_output --partial "v0.42"
+}
+
+# COMMANDS
+#
+# ck8s init
+# ck8s apply
+# ck8s upgrade prepare
+# ck8s upgrade apply
+#
+# ck8s init
+# next:
+#   ck8s apply
+#   ck8s upgrade prepare
+# not:
+#   ck8s upgrade apply
+#   ck8s init ?
+#
+# ck8s apply
+# next:
+#   ck8s apply
+#   ck8s upgrade prepare
+# not:
+#   ck8s upgrade apply
+#   ck8s init ?
+#
+# ck8s upgrade prepare
+# next:
+#   ck8s upgrade apply
+# not:
+#   ck8s apply
+#   ck8s init ?
+#
+# ck8s upgrade apply
+# next:
+#   ck8s apply ?
+#
+# any other combo must throw error
+
+@test "no upgrade apply without upgrade prepare" {
+  # test 2
+
+  gitversion.mock_static "v0.42.0"
+  run ck8s version config
+  assert_success
+  assert_output --partial "v0.41"
+  run ck8s upgrade sc "v0.42" apply
+  assert_failure # because trying to apply 0.42 when config says 0.41
+  assert_output --partial "invalid upgrade path"
+}
+
+@test "no ck8s apply without ck8s upgrade" {
+  # test 3
+
+  gitversion.mock_static "v0.42.0"
+  run ck8s upgrade sc "v0.42" prepare
+  assert_success
+  assert_output --partial "locked for upgrade"
+
+  run ck8s ops kubectl sc get -n kube-system configmap apps-upgrade -o jsonpath --template='{.data.version}'
+  assert_success
+  assert_output --partial "v0.42"
+
+  # Oops, forgot to ck8s upgrade apply
+
+  run ck8s apply sc
+  assert_failure
+  assert_output --partial "Upgrade ongoing"
+}
+
+@test "prevent apply using older config" {
+  # test 4
+  #
+  # pretend we already upgraded
+  run yq -i '.global.ck8sVersion="v0.42.0"' "${CK8S_CONFIG_PATH}/defaults/common-config.yaml"
+  ck8s ops kubectl sc apply -f - <<EOF
+apiVersion: v1
+data:
+  version: v0.41
+kind: ConfigMap
+metadata:
+  name: apps-meta
+  namespace: kube-system
+EOF
+
+  # downgrade
+  gitversion.mock_static "v0.41.0"
+  run ck8s apply sc
+  assert_failure
+  assert_output --partial "Version mismatch. Run upgrade to update config."
+}
```

Comment:
The test title is "prevent apply using older config" but it looks like it's testing when the config is newer (v0.42.0) and the cluster is older (v0.41)?

Reply:
I don't think the configmap patch makes any difference in that test. The error message you get is from the version in the config file mismatches from the git repository.

If you actually want to test the apply failing because of the diff in the configmap the configuration file needs to match v0.41 as well, no?

Reply:
I think this still needs to be addressed before I can approve.

Reply:
Hm, right. How would you write the test?